File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.java
import java.util.PriorityQueue;
class Solution {
    public int solution(int n, int k, int[] enemy) {
        int answer = 0;
        if (k >= enemy.length) {
            return enemy.length;
        }
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int i = 0; i < enemy.length; i++) {
            pq.offer(enemy[i]);
            if (pq.size() > k) {
                int tmp = pq.poll();
                if (n - tmp < 0) {
                    return i;
                }
                n -= tmp;
            }
        }
        return enemy.length;
    }
}

File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.py
from heapq import heappush, heappop
def solution(n, k, enemy):
    if k >= len(enemy):
        return len(enemy)
    
    q = []
    for i in range(len(enemy)):
        heappush(q, enemy[i])
        if len(q) > k:
            tmp = heappop(q)
            if n - tmp < 0:
                return i
            n -= tmp
    return len(enemy)